### ISSUE TITLE
Bug Report: Type narrowing not working as expected

### ISSUE THREAD

**Issue by antonio-ivanovski** [microsoft/typescript-go#912]
**Expected Behavior:**

When using type guards like `Array.isArray` or truthiness checks in an `if` block, the type of the variable within that block should be narrowed down to the more specific type.

For `Array.isArray(foo)`:
```typescript
const foo: string | string[] = 'foo'; // or ['foo']
if (Array.isArray(foo)) {
    // foo should be inferred as string[]
    foo; 
}
```

For truthiness check on `bar`:
```typescript
const bar: string | undefined = 'bar'; // or undefined
if (bar) {
    // bar should be inferred as string
    bar;
}
```

**Actual Behavior:**

The type of the variable inside the `if` block is not being narrowed correctly.

For `Array.isArray(foo)`:
```typescript
const foo: string | string[] = 'foo'; // or string[]
if (Array.isArray(foo)) {
    // foo is inferred as string | string[]
    foo; 
}
```
The type of `foo` remains `string | string[]` inside the `if` block, even after the `Array.isArray(foo)` check.

For truthiness check on `bar`:
```typescript
const bar: string | undefined = 'bar'; // or undefined
if (bar) {
    // bar is inferred as string | undefined
    bar.charAt(0); // Error: Object is possibly 'undefined'.
}
```
The type of `bar` remains `string | undefined` inside the `if` block, even after the truthiness check. This can lead to unnecessary errors or require explicit type assertions.

<img width="346" alt="Image" src="https://github.com/user-attachments/assets/722e2c71-a386-42d2-a82b-ddb35e9e84da" />

<img width="340" alt="Image" src="https://github.com/user-attachments/assets/92aebcf6-799a-4c55-9e3d-9ee10610ecdb" />

**Additional Context:**

@typescript/native-preview: "7.0.0-dev.20250522.2"
TypeScript (Native Preview): 0.20250522.2

**Comment by DanielRosenwasser** [comment]
I think this is a bug more in quick info than in narrowing itself - you can tell because an error is reported on the following:

```ts
function f(foo: string | string[]) {
    if (Array.isArray(foo)) {
        let x: string = foo; // error

        let y = foo; // hovering on 'y' shows 'string[]'
    }
}
```

Quick info is currently in development so I guess it's currently only showing the declared type

**Comment by antonio-ivanovski** [comment]
> I think this is a bug more in quick info than in narrowing itself - you can tell because an error is reported on the following:
> 
> ```ts
> function f(foo: string | string[]) {
>     if (Array.isArray(foo)) {
>         let x: string = foo; // error
> 
>         let y = foo; // hovering on 'y' shows 'string[]'
>     }
> }
> ```
> 
> Quick info is currently in development so I guess it's currently only showing the declared type

Haven't checked the particular case you mentioned, but in the case with `string | undefined`, I was not getting an error in VS Code when I was trying to access property on potentially undefined property. Not sure if this also falls under the category of quick info.

**Comment by DanielRosenwasser** [comment]
Is it possible you were running in a file without `--strict` on?

![Image](https://github.com/user-attachments/assets/a64c719c-be22-401a-8c5a-3e9e182388a7)

**Comment by antonio-ivanovski** [comment]
> Is it possible you were running in a file without `--strict` on?
> 
> ![Image](https://github.com/user-attachments/assets/a64c719c-be22-401a-8c5a-3e9e182388a7)

Will recheck this tomorrow first thing in the morning. Thanks for taking a look so far.

**Comment by antonio-ivanovski** [comment]
Tested again and it seems that only the quick info in vs code is showing invalid type, the actual type under the hood is correct. 

<img width="487" alt="Image" src="https://github.com/user-attachments/assets/f6b2e575-5e0e-41ec-b241-a2c2da819b0e" />

<img width="547" alt="Image" src="https://github.com/user-attachments/assets/3d336ecb-132a-4753-b61b-307a630a26f2" />
